Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,442 @@
+            {
+                if (!m_shapeCache.ValidateCache(m_BoundingShape2D, m_MaxOrthoSize, m_confinerBaker, 
+                    state.Lens.Aspect, out bool confinerStateChanged))
+                {
+                    return; // invalid path
+                }
+                
+                float frustumHeight = CalculateHalfFrustumHeight(state, vcam);
+                m_extra = GetExtraState<VcamExtraState>(vcam);
+                m_extra.m_vcamShapeCache.ValidateCache(m_confinerBaker, confinerStateChanged, frustumHeight, state.Lens.Orthographic, m_extra);
+
+                Vector3 displacement = ConfinePoint(state.CorrectedPosition, m_extra.m_vcamShapeCache.m_path, 
+                    m_shapeCache.m_localToWorldDelta, Vector3.zero);
+                // Remember the desired displacement for next frame
+                var prev = m_extra.m_previousDisplacement;
+                m_extra.m_previousDisplacement = displacement;
+
+                if (!VirtualCamera.PreviousStateIsValid || deltaTime < 0 || m_Damping <= 0)
+                    m_extra.m_dampedDisplacement = Vector3.zero;
+                else
+                {
+                    // If a big change from previous frame's desired displacement is detected, 
+                    // assume we are going around a corner and extract that difference for damping
+                    if (Vector2.Angle(prev, displacement) > m_cornerAngleTreshold)
+                        m_extra.m_dampedDisplacement += displacement - prev;
+
+                    m_extra.m_dampedDisplacement -= Damper.Damp(m_extra.m_dampedDisplacement, m_Damping, deltaTime);
+                    displacement -= m_extra.m_dampedDisplacement;
+                }
+                state.PositionCorrection += displacement;
+            }
+        }
+
+        /// <summary>
+        /// Calculates half frustum height for orthographic or perspective camera.
+        /// For more info on frustum height, see <see cref="docs.unity3d.com/Manual/FrustumSizeAtDistance.html"/> 
+        /// </summary>
+        /// <param name="state">CameraState for checking if Orthographic or Perspective</param>
+        /// <param name="vcam">vcam, to check its position</param>
+        /// <returns>Frustum height of the camera</returns>
+        private float CalculateHalfFrustumHeight(in CameraState state, in CinemachineVirtualCameraBase vcam)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = Mathf.Abs(state.Lens.OrthographicSize);
+            }
+            else
+            {
+                Quaternion inverseRotation = Quaternion.Inverse(m_BoundingShape2D.transform.rotation);
+                Vector3 planePosition = inverseRotation * m_BoundingShape2D.transform.position;
+                Vector3 cameraPosition = inverseRotation * vcam.transform.position;
+                float distance = Mathf.Abs(planePosition.z - cameraPosition.z);
```

Comment:
Why do you have to do all this to get the distance?  Why not just Vector3.Magnitude(boundingShape.position - camera.position)?

Reply:
The reason, why I did it this way, is that what if the collider and the camera is rotated. So that distance is not determined by the Z axis.
I am interested in the distance between the collider's plane and the camera.

However, I simplified it to just using Z distance, because rotating the collider and camera is weird.

Reply:
Why not just do a pure distance, as I suggested?  I think it's safe to assume that the camera is not looking obliquely at the collider (nothing would work in that case).  No need to assume that it's looking along the z-axis (although that is the most common case).  It's also safe to assume that the confiner shape is in a plane that passes through the confiner shape's transform position.